ouroboros-network/src/Ouroboros/Network/Subscription/Common.hs

```diff
@@ -0,0 +1,383 @@
+    -> Maybe Socket.SockAddr
+    -> Maybe Socket.SockAddr
+    -> (Socket.SockAddr -> Maybe DiffTime)
+    -> IPSubscriptionTarget
+    -> (Socket.Socket -> IO ())
+    -> (Async () -> IO t)
+    -> IO t
+ipSubscriptionWorker tbl tracer localIPv4 localIPv6 connectionAttemptDelay ips cb k =
+    subscriptionWorker
+            tbl
+            (ipListTracer localIPv4 localIPv6 (ispIps ips) tracer)
+            localIPv4
+            localIPv6
+            connectionAttemptDelay
+            (return $ listSubscriptionTarget (ispIps ips))
+            (ispValency ips)
+            cb k
+
+subscriptionWorker
+    :: ConnectionTable
+    -> Tracer IO SubscriptionTrace
+    -> Maybe Socket.SockAddr
+    -> Maybe Socket.SockAddr
+    -> (Socket.SockAddr -> Maybe DiffTime)
+    -> IO (SubscriptionTarget IO Socket.SockAddr)
+    -> Int
+    -> (Socket.Socket -> IO ())
+    -> (Async () -> IO t)
+    -> IO t
+subscriptionWorker tbl tracer localIPv4 localIPv6 connectionAttemptDelay getTargets valency
+        cb k = do
+    childrenVar <- newTVarM Set.empty
+    withAsync (worker childrenVar)
+              (\aid -> (k aid) `finally` (do
+                  children <- atomically $ readTVar childrenVar
+                  mapM_ (\a -> throwTo a
+                                   (SubscriberError SubscriberWorkerCancelled
+                                    "SubscriptionWorker exiting"
+                                    callStack)) $ Set.toList children
+              ))
+  where
+    worker :: TVar IO (Set ThreadId) -> IO ()
+    worker childrenVar = do
+        valencyVar <- newTVarM valency
+        traceWith tracer $ SubscriptionTraceStart valency
+        forever $ do
+            start <- getMonotonicTime
+            targets <- getTargets
+            subscribeTo tbl childrenVar tracer localIPv4 localIPv6 connectionAttemptDelay
+                    valencyVar cb targets
+            atomically $ do
+                v <- readTVar valencyVar
+                when (v <= 0)
+                    retry
+
+            {-
+            - We allways wait at least ipRetryDelay seconds between calls to getTargets, and
+            - before trying to restart the subscriptions we also wait 1 second so that if multiple
+            - subscription targets fail around the same time we will try to restart with a valency
+            - higher than 1.
+            -}
+            threadDelay 1
+            end <- getMonotonicTime
+            let duration = diffTime end start
+            currentValency <- atomically $ readTVar valencyVar
+            traceWith tracer $ SubscriptionTraceRestart duration currentValency
+                (valency - currentValency)
+
+            when (duration < ipRetryDelay) $
+                threadDelay $ ipRetryDelay - duration
+
+data WithIPList a = WithIPList {
+      wilIPv4  :: !(Maybe Socket.SockAddr)
+    , wilIPv6  :: !(Maybe Socket.SockAddr)
+    , wilDsts  :: ![Socket.SockAddr]
+    , wilEvent :: !a
```

Comment:
I'm guessing the first two are local, the next is the targets, and what is the last one?

Reply:
General `Control.Tracer` misery.